nibabel/cifti2/cifti2_axes.py

```diff
@@ -0,0 +1,1330 @@
+        affine : np.ndarray
+            (4, 4) array with the voxel to mm transformation (defaults to identity matrix)
+            Argument will be ignored for surface masks
+
+        Returns
+        -------
+        BrainModel which covers the provided mask
+        """
+        if affine is None:
+            affine = np.eye(4)
+        if np.asarray(affine).shape != (4, 4):
+            raise ValueError("Affine transformation should be a 4x4 array or None, not %r" % affine)
+        if mask.ndim == 1:
+            return cls.from_surface(np.where(mask != 0)[0], mask.size, name=name)
+        elif mask.ndim == 3:
+            voxels = np.array(np.where(mask != 0)).T
+            return cls(name, voxel=voxels, affine=affine, volume_shape=mask.shape)
+        else:
+            raise ValueError("Mask should be either 1-dimensional (for surfaces) or "
+                             "3-dimensional (for volumes), not %i-dimensional" % mask.ndim)
+
+    @classmethod
+    def from_surface(cls, vertices, nvertex, name='Cortex'):
+        """
+        Creates a new BrainModel axis describing the vertices on a surface
+
+        Parameters
+        ----------
+        vertices : np.ndarray
+            indices of the vertices on the surface
+        nvertex : int
+            total number of vertices on the surface
+        name : str
+            Name of the brain structure (e.g. 'CortexLeft' or 'CortexRight')
+
+        Returns
+        -------
+        BrainModel which covers (part of) the surface
+        """
+        cifti_name = cls.to_cifti_brain_structure_name(name)
+        return cls(cifti_name, vertex=vertices,
+                   nvertices={cifti_name: nvertex})
+
+    @classmethod
+    def from_mapping(cls, mim):
+        """
+        Creates a new BrainModel axis based on a CIFTI dataset
+
+        Parameters
+        ----------
+        mim : cifti2.Cifti2MatrixIndicesMap
+
+        Returns
+        -------
+        BrainModel
+        """
+        nbm = np.sum([bm.index_count for bm in mim.brain_models])
+        voxel = -np.ones((nbm, 3))
+        vertex = -np.ones(nbm)
+        name = []
+
+        nvertices = {}
+        affine, shape = None, None
+        for bm in mim.brain_models:
+            index_end = bm.index_offset + bm.index_count
+            is_surface = bm.model_type == 'CIFTI_MODEL_TYPE_SURFACE'
+            name.extend([bm.brain_structure] * bm.index_count)
+            if is_surface:
+                vertex[bm.index_offset: index_end] = bm.vertex_indices
+                nvertices[bm.brain_structure] = bm.surface_number_of_vertices
+            else:
+                voxel[bm.index_offset: index_end, :] = bm.voxel_indices_ijk
+                if affine is None:
+                    shape = mim.volume.volume_dimensions
+                    affine = mim.volume.transformation_matrix_voxel_indices_ijk_to_xyz.matrix
+                else:
```

Comment:
You're checking the same `mim.volume.volume_dimensions` and `mim.volume.transformation_matrix_voxel_indices_ijk_to_xyz.matrix` every loop through. I don't think these are going to change on iterations over `mim.brain_models` (unless the API is weirder than I think). You can probably just drop this `else` block altogether.

I'm assuming that the logic here is actually to avoid setting `affine` and `shape` unless there is a volume model in the `mim.brain_models`, and then using common volume information. If it's possible that there's differing volume information, I assume it will be within the `bm` object, not `mim`.

Reply:
I agree that there is no reason for the `else` block. I will get rid of it